From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JiaY-shi <shi05275@163.com>
Date: Mon, 31 Aug 2026 02:30:00 +0800
Subject: [PATCH] arm: mediatek: add GL.iNet GL-MT2500 target

Base the board description on OpenWrt's GL-MT2500 target and use the
environment offset from the vendor/OpenWrt GPT layout.  Migrate the legacy
OpenWrt boot command in memory so existing devices prefer Armbian on USB
without rewriting their persistent environment.

Signed-off-by: JiaY-shi <shi05275@163.com>

---
 arch/arm/dts/mt7981-glinet-gl-mt2500.dts  | 138 ++++++++++++++++++++++
 board/mediatek/mt7981/mt7981_rfb.c        |  24 ++++
 configs/mt7981_glinet_gl-mt2500_defconfig |  82 +++++++++++++
 3 files changed, 244 insertions(+)
 create mode 100644 arch/arm/dts/mt7981-glinet-gl-mt2500.dts
 create mode 100644 configs/mt7981_glinet_gl-mt2500_defconfig

diff --git a/arch/arm/dts/mt7981-glinet-gl-mt2500.dts b/arch/arm/dts/mt7981-glinet-gl-mt2500.dts
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/arch/arm/dts/mt7981-glinet-gl-mt2500.dts
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/dts-v1/;
+#include "mt7981.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/linux-event-codes.h>
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	model = "GL.iNet GL-MT2500";
+	compatible = "glinet,gl-mt2500", "mediatek,mt7981";
+
+	chosen {
+		stdout-path = &uart0;
+		tick-timer = &timer0;
+	};
+
+	memory@40000000 {
+		device_type = "memory";
+		reg = <0x40000000 0x40000000>;
+	};
+
+	reg_3p3v: regulator-3p3v {
+		compatible = "regulator-fixed";
+		regulator-name = "fixed-3.3V";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	usb_vbus: regulator-usb {
+		compatible = "regulator-fixed";
+		regulator-name = "usb-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpios = <&pio 12 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 1 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-vpn {
+			label = "white:vpn";
+			gpios = <&pio 31 GPIO_ACTIVE_LOW>;
+		};
+
+		white {
+			label = "white:system";
+			gpios = <&pio 30 GPIO_ACTIVE_LOW>;
+		};
+
+		blue {
+			label = "blue:system";
+			gpios = <&pio 29 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&eth {
+	status = "okay";
+	mediatek,gmac-id = <1>;
+	phy-mode = "gmii";
+	phy-handle = <&phy0>;
+
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		phy0: ethernet-phy@0 {
+			compatible = "ethernet-phy-id03a2.9461";
+			reg = <0>;
+			phy-mode = "gmii";
+		};
+	};
+};
+
+&mmc0 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc0_pins_default>;
+	bus-width = <8>;
+	max-frequency = <26000000>;
+	vmmc-supply = <&reg_3p3v>;
+	cap-mmc-hw-highspeed;
+	cap-mmc-hw-reset;
+	non-removable;
+};
+
+&pio {
+	mmc0_pins_default: mmc0-pins-default {
+		mux {
+			function = "flash";
+			groups = "emmc_45";
+		};
+
+		conf-cmd-dat {
+			pins = "SPI0_CLK", "SPI0_MOSI", "SPI0_MISO",
+			       "SPI0_CS", "SPI0_HOLD", "SPI0_WP",
+			       "SPI1_CLK", "SPI1_MOSI", "SPI1_MISO";
+			input-enable;
+			drive-strength = <MTK_DRIVE_12mA>;
+			bias-pull-up = <MTK_PUPD_SET_R1R0_01>;
+		};
+
+		conf-clk {
+			pins = "SPI1_CS";
+			drive-strength = <MTK_DRIVE_12mA>;
+			bias-pull-down = <MTK_PUPD_SET_R1R0_10>;
+		};
+
+		conf-rst {
+			pins = "GPIO_WPS";
+			drive-strength = <MTK_DRIVE_4mA>;
+			bias-pull-up = <MTK_PUPD_SET_R1R0_01>;
+		};
+	};
+};
+
+&uart0 {
+	status = "okay";
+};
+
+&xhci {
+	vbus-supply = <&usb_vbus>;
+};
diff --git a/board/mediatek/mt7981/mt7981_rfb.c b/board/mediatek/mt7981/mt7981_rfb.c
index 9c9cd657188e..222222222222 100644
--- a/board/mediatek/mt7981/mt7981_rfb.c
+++ b/board/mediatek/mt7981/mt7981_rfb.c
@@ -6,6 +6,7 @@
  */
 
 #include <config.h>
+#include <dm.h>
 #include <env.h>
 #include <init.h>
 #include <asm/global_data.h>
@@ -16,14 +17,37 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
+static void gl_mt2500_migrate_legacy_environment(void)
+{
+	const char *bootcmd = env_get("bootcmd");
+
+	if (bootcmd && strcmp(bootcmd, "run boot_system") &&
+	    strcmp(bootcmd, "run distro_bootcmd") &&
+	    strcmp(bootcmd, "bootflow scan -lb"))
+		return;
+
+	env_set("scriptaddr", "0x50000000");
+	env_set("kernel_addr_r", "0x52000000");
+	env_set("fdt_addr_r", "0x62000000");
+	env_set("ramdisk_addr_r", "0x62180000");
+	env_set("boot_openwrt", "mmc dev 0; gpt setenv mmc 0 kernel; "
+		"mmc read ${kernel_addr_r} ${gpt_partition_addr} "
+		"${gpt_partition_size}; bootm ${kernel_addr_r}");
+
+	env_set("bootcmd", "usb start; sysboot usb 0:5 any ${scriptaddr} "
+		"/boot/extlinux/extlinux.conf; run boot_openwrt");
+}
+
 int board_init(void)
 {
 	return 0;
 }
 
 int board_late_init(void)
 {
 	gd->env_valid = 1; //to load environment variable from persistent store
 	env_relocate();
+	if (of_machine_is_compatible("glinet,gl-mt2500"))
+		gl_mt2500_migrate_legacy_environment();
 	return 0;
 }
diff --git a/configs/mt7981_glinet_gl-mt2500_defconfig b/configs/mt7981_glinet_gl-mt2500_defconfig
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/configs/mt7981_glinet_gl-mt2500_defconfig
@@ -0,0 +1,86 @@
+CONFIG_ARM=y
+CONFIG_SYS_HAS_NONCACHED_MEMORY=y
+CONFIG_POSITION_INDEPENDENT=y
+CONFIG_ARCH_MEDIATEK=y
+CONFIG_TEXT_BASE=0x41e00000
+CONFIG_SYS_MALLOC_F_LEN=0x4000
+CONFIG_NR_DRAM_BANKS=1
+CONFIG_ENV_SIZE=0x80000
+CONFIG_ENV_OFFSET=0x400000
+CONFIG_DEFAULT_DEVICE_TREE="mt7981-glinet-gl-mt2500"
+CONFIG_TARGET_MT7981=y
+CONFIG_SYS_LOAD_ADDR=0x46000000
+CONFIG_DEBUG_UART_BASE=0x11002000
+CONFIG_DEBUG_UART_CLOCK=40000000
+CONFIG_DEBUG_UART=y
+# CONFIG_EFI_LOADER is not set
+CONFIG_FIT=y
+CONFIG_BOOTSTD_FULL=y
+CONFIG_BOOTDELAY=1
+CONFIG_AUTOBOOT_KEYED=y
+CONFIG_AUTOBOOT_FLUSH_STDIN=y
+CONFIG_AUTOBOOT_PROMPT="Autoboot in %d seconds, press <SPACE> to stop\n"
+CONFIG_AUTOBOOT_STOP_STR=" "
+CONFIG_DEFAULT_FDT_FILE="mediatek/mt7981b-glinet-gl-mt2500-v1.dtb"
+CONFIG_SYS_CBSIZE=512
+CONFIG_SYS_PBSIZE=1049
+CONFIG_LOGLEVEL=7
+CONFIG_LOG=y
+CONFIG_BOARD_LATE_INIT=y
+CONFIG_SYS_PROMPT="GL-MT2500> "
+# CONFIG_BOOTM_NETBSD is not set
+# CONFIG_BOOTM_PLAN9 is not set
+# CONFIG_BOOTM_RTEMS is not set
+# CONFIG_BOOTM_VXWORKS is not set
+# CONFIG_CMD_ELF is not set
+# CONFIG_CMD_UNLZ4 is not set
+# CONFIG_CMD_UNZIP is not set
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_GPT=y
+CONFIG_CMD_GPT_RENAME=y
+CONFIG_CMD_LSBLK=y
+CONFIG_CMD_BOOTMENU=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_READ=y
+CONFIG_CMD_WRITE=y
+CONFIG_CMD_USB=y
+CONFIG_CMD_EXT4=y
+CONFIG_CMD_EXT4_WRITE=y
+CONFIG_CMD_FS_UUID=y
+CONFIG_CMD_SYSBOOT=y
+CONFIG_CMD_PING=y
+CONFIG_CMD_WGET=y
+CONFIG_CMD_HASH=y
+CONFIG_CMD_SMC=y
+CONFIG_CMD_FAT=y
+CONFIG_CMD_FS_GENERIC=y
+CONFIG_PARTITION_TYPE_GUID=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_MMC=y
+CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_REGMAP=y
+CONFIG_SYSCON=y
+CONFIG_CLK=y
+CONFIG_MMC_HS200_SUPPORT=y
+CONFIG_MMC_MTK=y
+CONFIG_PHY_FIXED=y
+CONFIG_MEDIATEK_ETH=y
+CONFIG_PHY=y
+CONFIG_PHY_MTK_TPHY=y
+CONFIG_USB=y
+CONFIG_USB_XHCI_HCD=y
+CONFIG_USB_XHCI_MTK=y
+CONFIG_USB_STORAGE=y
+CONFIG_PINCTRL=y
+CONFIG_PINCONF=y
+CONFIG_PINCTRL_MT7981=y
+CONFIG_POWER_DOMAIN=y
+CONFIG_MTK_POWER_DOMAIN=y
+CONFIG_DM_REGULATOR=y
+CONFIG_DM_REGULATOR_FIXED=y
+CONFIG_DM_SERIAL=y
+CONFIG_MTK_SERIAL=y
+CONFIG_FAT_WRITE=y
+CONFIG_HEXDUMP=y
